spring-cloud-stream-binder-rabbit-core/src/main/java/org/springframework/cloud/stream/binder/rabbit/provisioning/RabbitExchangeQueueProvisioner.java

```diff
@@ -0,0 +1,490 @@
+	private Exchange buildExchange(RabbitCommonProperties properties, String exchangeName) {
+		try {
+			// TODO Make the ctor public in Spring-AMQP - AMQP-695
+			Constructor<ExchangeBuilder> ctor = ExchangeBuilder.class.getDeclaredConstructor(String.class, String.class);
+			ReflectionUtils.makeAccessible(ctor);
+			ExchangeBuilder builder = ctor.newInstance(exchangeName, properties.getExchangeType());
+			if (properties.isDelayedExchange()) {
+				builder.delayed();
+			}
+			return builder.build();
+		}
+		catch (Exception e) {
+			throw new IllegalStateException("Failed to create exchange object", e);
+		}
+	}
+
+	private void declareExchange(final String rootName, final Exchange exchange) {
+		try {
+			this.rabbitAdmin.declareExchange(exchange);
+		}
+		catch (AmqpConnectException e) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug(
+						"Declaration of exchange: " + exchange.getName() + " deferred - connection not available");
+			}
+		}
+		addToAutoDeclareContext(rootName + ".exchange", exchange);
+	}
+
+	private void addToAutoDeclareContext(String name, Object bean) {
+		synchronized (this.autoDeclareContext) {
+			if (!this.autoDeclareContext.containsBean(name)) {
+				this.autoDeclareContext.getBeanFactory().registerSingleton(name, bean);
+			}
+		}
+	}
+
+	private void declareBinding(String rootName, org.springframework.amqp.core.Binding binding) {
+		try {
+			this.rabbitAdmin.declareBinding(binding);
+		}
+		catch (AmqpConnectException e) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug(
+						"Declaration of binding: " + rootName + ".binding deferred - connection not available");
+			}
+		}
+		addToAutoDeclareContext(rootName + ".binding", binding);
+	}
+
+	public void cleanAutoDeclareContext(String name) {
+		synchronized (this.autoDeclareContext) {
+			removeSingleton(name + ".binding");
+			removeSingleton(name);
+			String dlq = name + ".dlq";
+			removeSingleton(dlq + ".binding");
+			removeSingleton(dlq);
+		}
+	}
+
+	private void removeSingleton(String name) {
+		if (this.autoDeclareContext.containsBean(name)) {
+			ConfigurableListableBeanFactory beanFactory = this.autoDeclareContext.getBeanFactory();
+			if (beanFactory instanceof DefaultListableBeanFactory) {
+				((DefaultListableBeanFactory) beanFactory).destroySingleton(name);
+			}
+		}
+	}
+
+	private final class RabbitProducerDestination implements ProducerDestination {
+
+		private final Exchange exchange;
+
+		private RabbitProducerDestination(Exchange exchange) {
+			this.exchange = exchange;
+		}
+
+		@Override
+		public String getProducerDestinationName() {
+			return this.exchange.getName();
+		}
+
+		@Override
+		public String getPartitionedProducerDestinationName(int partition) {
+			return this.exchange.getName();
+		}
+
+		@Override
+		public String toString() {
+			return this.exchange.toString();
+		}
+	}
+
+	private final class RabbitConsumerDestination implements ConsumerDestination {
+
+		private final Queue queue;
+
```

Comment:
We should hold a reference to the binding too and include it in the `toString()`.